### DISCUSSION TITLE
Controlling source port for `initrd`

### DISCUSSION THREAD

**Discussion by upsj** [ipxe/ipxe#1282] [Q&A] [answered]
Hi all, I am using iPXE with warewulf for stateless provisioning of nodes. Similar to `nfs`, warewulf uses a `secure` configuration option that controls access to certain files used during provisioning, requiring a privileged source port to restrict access to privileged users in the internal network. It seems that this functionality is currently not actually working, because there isn't an obvious way to tell `initrd`/`imgextract` to use a privileged source port for the HTTP fetch. Are there any ways to make this work, or is it possible to extend iPXE to allow for this? A short look through the code lets me assume that the source port is not set explicitly anywhere.

**Comment by mcb30** [comment, marked as answer]
No, there's no way to control the source port.  The source port is an arbitrary choice of the client and anything that relies on the source port number for security is broken by design, so we won't be changing this absent a very compelling reason to do so.

**Reply by upsj** [reply]
Thanks for the quick response!

> anything that relies on the source port number for security is broken by design

I don't think that is true in a private network where all systems are known (which is also why nfs uses it to differentiate between mounts by `root` or other users on a node), but I get your point.

**Comment by NiKiZe** [comment]
"privileged source port" is a concept that only applies to some TCP stacks, escentially creating more of a "security by obscurity" and not really secure in any way.

**Reply by upsj** [reply]
Not trying to argue against the concern here, but it applies to the Linux TCP stack and can thus create a trusted environment if you control all nodes in the network, which is not uncommon for provisioning. warewulf is mostly used for high performance computing clusters, where all compute nodes share the same internal network on a separate switch, so the source port is a valid approach to distinguish between privileged and non-privileged processes.

**Reply by NiKiZe** [reply]
How possible the security might be in this specific setup is irrelevant since it in most cases are invalid, and the feature in itself sets an insecure precedent. It will probably raise one or more CVEs

**Reply by mcb30** [reply]
> Not trying to argue against the concern here, but it applies to the Linux TCP stack and can thus create a trusted environment if you control all nodes in the network, which is not uncommon for provisioning. warewulf is mostly used for high performance computing clusters, where all compute nodes share the same internal network on a separate switch, so the source port is a valid approach to distinguish between privileged and non-privileged processes.

If you really want this level of security (i.e. all nodes implicitly trust all other nodes to enforce a network-wide concept of privilege), then use separate VLANs and assume each VLAN is a fully trustworthy security domain with no need to care about source port numbers.  Adding a VLAN interface under Linux requires essentially the same level of local privilege as binding to a port below 1024.

This approach has the advantage that you can in future upgrade to e.g. 802.1x port authentication to control access to the VLANs, which gives you a path towards adding some level of protection against unauthorised nodes.
